crates/prune/src/pruner.rs

```diff
@@ -114,172 +98,102 @@ impl<DB: Database> Pruner<DB> {
                 .map_or(1, |previous_tip_block_number| tip_block_number - previous_tip_block_number)
                 as usize;
 
-        if let (Some((to_block, prune_mode)), true) =
-            (self.modes.prune_target_block_receipts(tip_block_number)?, delete_limit > 0)
-        {
-            trace!(
-                target: "pruner",
-                segment = ?PruneSegment::Receipts,
-                %to_block,
-                ?prune_mode,
-                "Got target block to prune"
-            );
-
-            let segment_start = Instant::now();
-            let segment = segments::Receipts::default();
-            let output = segment.prune(&provider, PruneInput { to_block, delete_limit })?;
-            if let Some(checkpoint) = output.checkpoint {
-                segment.save_checkpoint(&provider, checkpoint.as_prune_checkpoint(prune_mode))?;
+        // TODO(alexey): this is cursed, refactor
+        #[allow(clippy::type_complexity)]
+        let segments: [(
+            Box<dyn Segment<DB>>,
+            Box<
+                dyn Fn(
+                    &PruneModes,
+                    BlockNumber,
+                )
+                    -> Result<Option<(BlockNumber, PruneMode)>, PruneSegmentError>,
+            >,
+        ); 5] = [
```

Comment:
I think we can at least add a helper type for this, with the segment and the handler as field,

or maybe the fn can even be made part of the segment trait?

Reply:
> I think we can at least add a helper type for this, with the segment and the handler as field,

Yep, introduced a helper type. Couldn't make it more succinct because of `Fn` definition.

> or maybe the fn can even be made part of the segment trait?

I think it can be, but it would also require touching `PruneModes`, and I rather keep this PR only to refactoring the pruner, because it's already large.